namespace SharedService.SharedKernel.Errors;

public static class Errors
{
    public static class General
    {
        public static Error ValueIsInvalid(string? name = null)
        {
            var label = name ?? "value";
            return Error.Validation(label, $"{label} is invalid");
        }

        public static Error NotFound(Guid? id = null)
        {
            var forId = id == null ?
                string.Empty :
                $"for id '{id}'";

            return Error.NotFound(
                "record.not.found",
                $"record not found {forId}");
        }

        public static Error AlreadyExists(string? name = null)
        {
            var label = name == null ?
                string.Empty :
                " " + name + " ";

            return Error.Validation(
                "record.already.exist",
                $"record with name: {label} already exist");
        }

        public static Error Failure(string? name = null)
        {
            var label = name == null ?
                string.Empty :
                " " + name + " ";

            return Error.Failure(label, $"{label} is invalid");
        }

        public static Error ValueIsRequired(string? name = null)
        {
            var label = name == null ?
                string.Empty :
                " " + name + " ";

            return Error.Validation(
                "lenght.is.invalid",
                $"invalid {label} lenght");
        }

        public static Error DeleteConflict(
            Guid? id = null, string? entityTypeName = null)
        {
            var forId = id is null ?
                " " :
                $"with id {id} ";

            var type = entityTypeName is null ?
                string.Empty :
                $"of type {entityTypeName}";

            return Error.Conflict("Conflict.Constraint", $"Can't delete entity {forId} {type}");
        }
    }
}